### PR TITLE
Update gh actions

### PR DESCRIPTION
**Pull request by mauwii**
Dios mio - it was such a pain in the ... to get this working, but damn does it feel good now where it worked out ^^

The test-invoke-conda.yml is now working with latest ubuntu and macOS, also activating the conda environments and attaching the generated pictures for each OS. It would be a good Idea to set your main branch protection to only allow merging pull requests if this action is successful, since it is confirming the usability on a clean system (so I would even go so far and say people with problems have one of two problems: 1) incompatible machine or 2) some dirty setup (like messed up libraries, packages for wrong architecture, ....)

My next try is to include windows in the os matrix as well, but since I almost never have been using windows for CI/CD stuff, I did not want to wait with the PR till it is included, since I even don't know if it will work out or not.

@lstein: I tried to merge this with the dev-branch, but got merge conflicts :-/ But since I do not have Merge-Conflicts when merging it into main and it also did not change code but confirm it's functionality, I thought it would be a good Idea to directly merge it into main. I hope you see this similar 🙈 

The mkdocs-material action does now deploy the whole docs without breaking emojis and stuff 🤓 It also runs when pushing/Pull requesting main- or development branch, but only deploys when running from refs/heads/main, so imho this could also be included to the main-branch-protection

Hope you like this, have a good one!

### CODE REVIEW

**Comment by mauwii** [comment]
Ah, 4got to clarify why I unpinned all versions from the packages in environment-mac.yml:

first of all --> this was the only solution to get it working, but also one of the functions of conda is to resolve conflicts between packages and another function is the possibilty of updating a environment, which both of em is of course not possible while every version is pinned. The [Conda Documentation](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-pkgs.html#preventing-packages-from-updating-pinning) also mentions to use another file to pin versions if it is necesarry to prevent a package from updateing...

So imho this was the best solution for this problem (and it should even solve the problem of people with macs who cant build the environment). And I think there is no better proof than the succeeding workflow ;P

**Comment by mauwii** [comment]
Maybe I already managed to integrate a windows-runner, but [unfortunatelly it fails](https://github.com/mauwii/stable-diffusion/actions/runs/3259500122/jobs/5352371313) by having not enough RAM 😕

**Comment by lstein** [comment]
I'm so glad that someone has done this. I was afraid to touch these files! I'm OK with it, but would like to see confirmation from the other reviewers before I go ahead and merge into main.

**Review by lstein — APPROVED** [review]
Insofar as I understand how the workflows work, this looks good to me.

**Review by lstein — CHANGES_REQUESTED** [review]
Please rebase your PR to merge the 122 commits into a single one. To do this at the command line, find the last commit from **before** your first commit, and run `git rebase <last commit>`

**Comment by mauwii** [comment]
> Please rebase your PR to merge the 122 commits into a single one. To do this at the command line, find the last commit from **before** your first commit, and run `git rebase <last commit>`

Are you sure? Since there are also commits of other people in between. Wouldn't it be easier if you squash-merge it onto the main?

**Comment by tildebyte** [comment]
@lstein I think in this instance we're OK to squash merge

**Comment by mauwii** [comment]
> @lstein I think in this instance we're OK to squash merge

Thx - there should even be no merge conflicts by doing so, while to rebase it with all the other commits in between creates a conflict on my side.

**Comment by tildebyte** [comment]
@mauwii;

> I unpinned all versions from the packages in environment-mac.yml

Please do so for environment.yml as well

Also, 🌮, this is something which really needed doing. I plan to follow up the installer work which cmdr2 is doing with something which will unpin everything on dev, and have everything properly pinned on main for releases ([pip compile](https://pip-tools.readthedocs.io/en/latest/), if you want to take a look)

**Comment by mauwii** [comment]
I do not get the point of pinning and using anaconda, couldn't you then instead not just use requirements.txt and venv (both already available in python and not creating needs to install conda or anything)?

So I didn't want to do something with environment.yml since it is actually building, but ofc could look into this as well.

Planned to look into the docker things since I learned a lot about kubernetes and want to grow knowledge in that direction, so if this also plays into your cards it would be nice 👯

**Comment by tildebyte** [comment]
@mauwii;

> you cannot test CI/CD stuff locally

Right, so what you do is open a PR against *your* fork, and directly merge into *your* 'main'. Once everything works as you want it to, you clean up your branch and open a PR against the upstream repo's 'main' (don't forget to `git reset --hard` *your* 'main'!)

File: .github/workflows/create-caches.yml [hunk 1]
```diff
@@ -29,42 +46,62 @@ jobs:
         with:
           path: models/ldm/stable-diffusion-v1/model.ckpt
           key: ${{ env.cache-name }}
-          restore-keys: |
-            ${{ env.cache-name }}
+          restore-keys: ${{ env.cache-name }}
+
+      - name: Use Cached Stable Diffusion v1.4 Model
```
Comment (tildebyte) [inline .github/workflows/create-caches.yml:51]: During my (unfortunately very lengthy and time-consuming 🙄) testing when *I* was trying to get Windows working, without success, I noticed that on average it's *much* faster to download the model every time ***because*** when you stash something in the lookaside cache to reuse, it gets zipped.

So instead of a straightforward download from Hugging Face's servers, there's a much more complicated download/unzip/move/not sure what else which *usually* takes longer (and yeah, this is across all three platforms).

File: .github/workflows/test-invoke-conda.yml [hunk 2]
```diff
@@ -4,29 +4,47 @@ on:
     branches:
       - 'main'
       - 'development'
+  pull_request:
```
Comment (tildebyte) [inline .github/workflows/test-invoke-conda.yml:7]: We don't want to do this. This will run the full suite on every push to a PR branch. The full suite is **extremely** time-consuming and ultimately not necessary anyway, as the existing config runs on *merge* to 'dev'.

The correct fix here would be to have some *minimal* test of some kind, but what we have is already as minimal as possible...

File: .github/workflows/create-caches.yml (hunk 1 above)
Comment (mauwii) [reply .github/workflows/create-caches.yml:51]: To be honest - I didn't find out how I need to create the secret with a token in a way that it is valid as long as the token - I cheated on my side with this secret (while still obtaining it from the huggingface site), but this was about 50% the speed I get from downloading it out of the cache (and the cache would still be valid after the secret has already lost it's authorization 🙈)

File: .github/workflows/test-invoke-conda.yml (hunk 2 above)
Comment (mauwii) [reply .github/workflows/test-invoke-conda.yml:7]: So should I remove the pull-request trigger? I mean, than this does not help anybody, if not preventing from merging non working PRs (since there is no package built/deployed/published/you-name it 🙈)

I can remove the PR Trigger 4 sure, but would think about this twice :-o

**Review by tildebyte — APPROVED** [review]

File: .github/workflows/create-caches.yml (hunk 1 above)
Comment (tildebyte) [reply .github/workflows/create-caches.yml:51]: Right, I forgot about that - I had to set up *my* token for it to work properly. The main repo should never have this issue, and if it does, it's something which we'd need to fix immediately

File: .github/workflows/test-invoke-conda.yml (hunk 2 above)
Comment (tildebyte) [reply .github/workflows/test-invoke-conda.yml:7]: Right, I get what you're saying, and on a normal project I'd agree with you, BUT:

Since the testing takes a REALLY long time, and triggers on *every push*, there are some PRs in here where the full test suite would be running 10s of times, which is not acceptable.

The "insurance" which we *do* have is that testing runs on *merges* to 'development'. That's not a perfect solution, but we've agreed as a project that it's OK for 'development' to be broken at times, as long as we make our best effort to keep it unbroken for the most part.

File: .github/workflows/create-caches.yml (hunk 1 above)
Comment (mauwii) [reply .github/workflows/create-caches.yml:51]: So should I disable the caching and always download, r u sure? imho this can be rly helpfull when there are many runs and saves bandwith of the backbone for people watching cat videos on YT ;P

File: .github/workflows/test-invoke-conda.yml (hunk 2 above)
Comment (mauwii) [reply .github/workflows/test-invoke-conda.yml:7]: Problem is that it would not work on peoples repositorys unless they set up the secret for the checkpoint themselfs, otherwise it could be done by first checking their branch

**Comment by tildebyte** [comment]
@mauwii;

> I do not get the point of pinning and using anaconda

I agree with you entirely about **anaconda**, but the best practice is to keep all dependencies un-pinned during development, then as the absolute last action before release, *after* the final merge to 'main', you run a utility to pin *every dependency* (and these days every dependency should be hashed as well), and then that's your release.

It doesn't make sense (for a **lot** of reasons) to *release* with unpinned dependencies; one big one for us is simply that no environment is reproducible at that point, because pip (which IIUC is also conda's resolver) will resolve all dependencies **on every install**. Given the pace of modern software development, you WILL be in  a situation where someone gets a working install one day, wipes it (for whatever reason) the next day, and installs again, only to find that their environment is completely broken because the dependencies can no longer be resolved.

TL;DR the burden of resolving (and pinning!) all dependencies to a known state is *on us*, and should be done at the end of a development cycle.

File: .github/workflows/test-invoke-conda.yml (hunk 2 above)
Comment (tildebyte) [reply .github/workflows/test-invoke-conda.yml:7]: Right, but we don't care about that, because we don't want to run on PRs

File: .github/workflows/create-caches.yml (hunk 1 above)
Comment (tildebyte) [reply .github/workflows/create-caches.yml:51]: > imho this can be rly helpfull when there are many runs

That's what I'm saying - **it isn't**. I was absolutely stunned by this at first, so I went down a total rabbithole of stopping work on getting Windows working just to focus on the model download.

ON AVERAGE, downloading the model is faster (by a lot) than caching it.

OTOH, "saves bandwith of the backbone" is another reason why we want to limit the number of times we run the suite.

File: .github/workflows/test-invoke-conda.yml (hunk 2 above)
Comment (mauwii) [reply .github/workflows/test-invoke-conda.yml:7]: So I take it out now, but imho I dont get the point of those Actions when only be running after somehing has been merged (imho it's one step to late for effectivley using those actions 🤷🏻‍♂️)

But np ;P

**Comment by tildebyte** [comment]
@mauwii; I don't know why I didn't just link to this in the first place https://medium.com/packagr/using-pip-compile-to-manage-dependencies-in-your-python-packages-8451b21a949e

Even if you only read the "The problem" section, it should be clear why pinning releases is best practice

File: .github/workflows/test-invoke-conda.yml (hunk 2 above)
Comment (tildebyte) [reply .github/workflows/test-invoke-conda.yml:7]: Because on 'development' we can immediately see if something is broken and fix it, and 'development' MUST NOT be merged to 'main' unless it is green, it which case we (should be) guaranteed that 'main' will pass as well.

**Comment by mauwii** [comment]
not sure why I cannot answer to your last comment @tildebyte, but just to mention it: I am very aware of CI/CD since working as DevOps since some Years 🤓 

I just ment that it would be easier in release then instead of using Conda to stick with requirements.txt since nobody needs to install anything besides a compatible version of python and can then

```bash
python -m venv .venv
source .venv/bin/activate
pip install -r requirements.txt
```

Why a production release should have pinned requirements is pretty obvious - but it would f.e. also differ between X86, ARM, and else (which is not done by the current environment yamls and therefore causes trouble sinc one party always has problems as it seems)

But I am totally with you about version pining for releases !

**Comment by tildebyte** [comment]
We're actually agreeing, but I'm not communicating effectively 😁 

```
python -m venv .venv
source .venv/bin/activate
pip install -r requirements.txt
```

There's a step *before* that (which should happen as the absolutely last thing done before release), which generates a fully-resolved, pinned, hashed 'requirements.txt' which *should* (given some luck) work on all platforms (pip hashes *all* of the wheels which it finds for a given package, but sometimes things don't work as they should :D)

So my proposal will be (in agreement with you) to completely drop Conda and replace it with 'pip compile'/'pip'/'venv', which currently (still a work in progress) looks like (this would be in a script)

```bash
# Commented lines are development-time only - end users should never run these. Also note
# that 'pip-tools' IS NOT an end-user dependency
# pip install pip-tools
# pip-compile -r --annotate --annotation-style split --generate-hashes --emit-index-url --allow-unsafe --reuse-hashes requirements.in
python -m venv invokeai
# IMO it will be easier for users to remember this, and we avoid any possibility of collision
source invokeai/Scripts/activate
python -m pip install --upgrade pip
pip install -r requirements.txt
# A source directory cannot be hashed, so pip will refuse to install it from a
# hashed 'requirements.txt' - it must be done "manually"
pip install -e .
```

File: .github/workflows/test-invoke-conda.yml (hunk 2 above)
Comment (mauwii) [reply .github/workflows/test-invoke-conda.yml:7]: And to not mess it up now: I should remove those lines, right:

```yaml
  pull_request:
    branches:
      - 'main'
      - 'development'
```

Which corresponds to lines 7-10

**Comment by mauwii** [comment]
@tildebyte: Yes, I just tried to point out that we aggre in why it should be pinned, but what is not clear to me in the end: Where would you get the different requirement files from (for Linux, Windows, Mac, and both in at least ARM and x64 imho). to in the end have something like

- requirements-linux-arm.txt
- requirements-linux-x64.txt (or amd64)
- requirements-macos-arm.txt
- ...

since you will be out of luck in having one allmighty requirements.txt

I think this should also be done with GH Actions but I am not sure if they have ARM runners available for free, since I am usually working on Azure-DevOps (fun fact: I just learned when using a macOS agent in DevOps it actually reffers to GitHub Runners xD )

File: .github/workflows/test-invoke-conda.yml (hunk 2 above)
Comment (mauwii) [reply .github/workflows/test-invoke-conda.yml:7]: Or do you mean Lines 3-6:

```yaml
  push:
    branches:
      - 'main'
      - 'development'
````

Comment (mauwii) [reply .github/workflows/test-invoke-conda.yml:7]: Ok, Io am pretty sure you mean 7-10

File: .github/workflows/create-caches.yml (hunk 1 above)
Comment (mauwii) [reply .github/workflows/create-caches.yml:51]: So for me the download from huggingface usually had betwen 50-100MB/s while GH-Cache had about 150MB/s, which is also handled internally and not over the web (where I am not sure if downloading from huggingface is maybe slower because it needs to be routed externally), which would lead to --> using cache doesn't bother the backend at all ;P

**Comment by tildebyte** [comment]
> Where would you get the different requirement files from

It shouldn't be necessary to have them, because "pip hashes *all* of the wheels which it finds for a given package" - but it's true that it doesn't always work out in practice. We might have to rely on having a contributor with e.g. ARM hardware who could generate the 'requirements.txt' for us. I think, at least to get started, we *should* generate a separate 'requirements.txt' for each platform (OS + hardware) *and then diff them*

Basically, [this](https://pip-tools.readthedocs.io/en/latest/#cross-environment-usage-of-requirements-in-requirements-txt-and-pip-compile)

If we are forced to go the route of per-platform requirements files, there's a spec for that (wish I'd known sooner) e.g. 'win32-py3.7-requirements.txt', 'macos-py3.10-requirements.txt'

**Comment by mauwii** [comment]
just to mention it again: You are already using a Conda environment with pinned version, one for mac and one for linux.

Since macOS is imho the hugest one regarding Users with ARM CPUs, it is already pointing out that one of both fractions always has problems to install the conda environment, then new versions are pinned in the environment, then somehow the other half is starting to get errors.

I have a Azure DevOps Instance with 10 Agents available (where I also can run on ARM or on amd64), and could of course create a project for Invoke-AI in it, but it would then maybe make more sence to "pull" me in the organisation so that we can directly do this with the main repo and not automate something like main -> fork -> main

BEsides that it would also be possible to achieve something with Docker where you could use Moby to build f.E. a ARM64 container on x64 Hardware, install the Conda Env, Freeze the PIP-State, export it, ....

**Comment by tildebyte** [comment]
> You are already using a Conda environment with pinned version, one for mac and one for linux

I *might* be misunderstanding what you're trying to say here, but from what I can see, the problem is *Conda* , not pinning, especially given the fact that we have *a separate file just for Mac*

People who are directly creating venvs, or (me) using pipenv, and then installing directly via 'pip', have *far* fewer problems thann people using Conda on *any* environment

**Comment by mauwii** [comment]
No, Conda solves a problem here, since it is able to resolve conflicts between packages by itself. But this ability is drawn away from Conta when conseqently pinning all packages to its strict version (instead f.E. using packagename==1.2.* or packagename>=1.0, <2.0)
